lib/db_connection/connection.ex

```diff
@@ -563,10 +568,17 @@ defmodule DBConnection.Connection do
     end
   end
 
-  defp handle_timeout(%{client: nil, idle_timeout: idle_timeout} = s) do
-    {:noreply, s, idle_timeout}
+  defp handle_timeout(%{client: nil, idle_timeout: idle_timeout,
+                        idle_hibernate: idle_hibernate} = s) do
+    Process.send_after(self(), :timeout, idle_timeout)
```

Comment:
If we use a timer then we need to be careful to cancel the timer and/or have the timeout message uniquely identifiable (include reference) because the `:timeout` could arrive late and trigger early timeout and out of sync timeouts.

This module includes few examples of `:erlang.start_timer` to do this in similar way. Given timers have better behavior here (timers give stricter timeout than a timeout return value) its a good change. This might complicate code, I think that might be why didn't implemented like that originally.

Seems like we really need `gen_statem` here but cant until otp 20+.